packages/remote/src/electron-node/ssh/remote-ssh-connection-provider.ts

```diff
@@ -0,0 +1,355 @@
+                } catch (err) {
+                    deferred.reject(err);
+                }
+            }).on('error', err => {
+                deferred.reject(err);
+            }).connect({
+                host: host,
+                username: user,
+                authHandler: (methodsLeft, successes, callback) => (sshAuthHandler(methodsLeft, successes, callback), undefined)
+            });
+        return deferred.promise;
+    }
+
+    /**
+     * Sometimes, ssh2.exec will not execute and retrieve any data right after the `ready` event fired.
+     * In this method, we just perform `echo hello` in a loop to ensure that the connection is really ready.
+     * See also https://github.com/mscdex/ssh2/issues/48
+     */
+    protected async testConnection(connection: RemoteSSHConnection): Promise<void> {
+        for (let i = 0; i < 100; i++) {
+            const result = await connection.exec('echo hello');
+            if (result.stdout.includes('hello')) {
+                return;
+            }
+            await timeout(50);
+        }
+        throw new Error('SSH connection failed testing. Could not execute "echo"');
+    }
+
+    protected getAuthHandler(user: string, host: string, identityKeys: SSHKey[]): ssh2.AuthHandlerMiddleware {
+        let passwordRetryCount = this.passwordRetryCount;
+        let keyboardRetryCount = this.passphraseRetryCount;
+        // `false` is a valid return value, indicating that the authentication has failed
+        const END_AUTH = false as unknown as ssh2.AuthenticationType;
+        // `null` indicates that we just want to continue with the next auth type
+        // eslint-disable-next-line no-null/no-null
+        const NEXT_AUTH = null as unknown as ssh2.AuthenticationType;
+        return async (methodsLeft: string[] | null, _partialSuccess: boolean | null, callback: ssh2.NextAuthHandler) => {
+            if (!methodsLeft) {
+                return callback({
+                    type: 'none',
+                    username: user,
+                });
+            }
+            if (methodsLeft && methodsLeft.includes('publickey') && identityKeys.length) {
+                const identityKey = identityKeys.shift()!;
+                if (identityKey.isPrivate) {
+                    return callback({
+                        type: 'publickey',
+                        username: user,
+                        key: identityKey.parsedKey
+                    });
+                }
+                if (!await fs.pathExists(identityKey.filename)) {
+                    // Try next identity file
+                    return callback(NEXT_AUTH);
+                }
+
+                const keyBuffer = await fs.promises.readFile(identityKey.filename);
+                let result = ssh2.utils.parseKey(keyBuffer); // First try without passphrase
+                if (result instanceof Error && result.message === 'Encrypted private OpenSSH key detected, but no passphrase given') {
```

Comment:
How stable is this error message?

Reply:
Good question - I've made this a bit more flexible only listening on the `no passphrase given` info, which seems to be consistent across all ssh client implementations.

Reply:
But does it make sense to even check the error message? Can't we just try _with_ the passphrase if it doesn't work without in any case?